File: src/main/java/com/github/tammo/fun/frontend/type/SubstitutionApplier.java
package com.github.tammo.fun.frontend.type;

import com.github.tammo.fun.frontend.type.TypedTreeNode.ArithmeticExpression;
import com.github.tammo.fun.frontend.type.TypedTreeNode.BinaryArithmeticExpression;
import com.github.tammo.fun.frontend.type.TypedTreeNode.NotImplemented;
import com.github.tammo.fun.frontend.type.TypedTreeNode.ParenthesizedExpression;

public class SubstitutionApplier {

    private final Unifier unifier;

    public SubstitutionApplier(Unifier unifier) {
        this.unifier = unifier;
    }

    public TypedTreeNode apply(TypedTreeNode node) {
        if (node == null) {
            return null;
        }

        // Rebuild the node based on its specific subtype
        return switch (node) {
            // ------------------------------------------------
            // Example: CompilationUnit
            // ------------------------------------------------
            case TypedTreeNode.CompilationUnit cu ->
                    new TypedTreeNode.CompilationUnit(
                            cu.namespaceDeclaration(),
                            (TypedTreeNode.ClassDeclaration) apply(cu.classDeclaration())
                    );

            // ------------------------------------------------
            // Example: ClassDeclaration
            // ------------------------------------------------
            case TypedTreeNode.ClassDeclaration cd -> {
                var newEffects = cd.effects().stream()
                        .map(this::apply)
                        .map(TypedTreeNode.EffectDeclaration.class::cast)
                        .toList();

                var newFunctions = cd.functions().stream()
                        .map(this::apply)
                        .map(TypedTreeNode.FunctionDeclaration.class::cast)
                        .toList();

                // Build the new ClassDeclaration
                yield new TypedTreeNode.ClassDeclaration(cd.name(), newEffects, newFunctions);
            }

            // ------------------------------------------------
            // Example: EffectDeclaration
            // ------------------------------------------------
            case TypedTreeNode.EffectDeclaration ed -> {
                TypedTreeNode.TypedIdentifier newTypedIdentifier = (TypedTreeNode.TypedIdentifier) apply(ed.name());

                var newParams = ed.parameters().stream()
                        .map(this::apply)
                        .map(TypedTreeNode.Parameter.class::cast)
                        .toList();

                TypedTreeNode.Expression newBody = (TypedTreeNode.Expression) apply(ed.body());

                yield new TypedTreeNode.EffectDeclaration(
                        newTypedIdentifier,
                        newParams,
                        newBody
                );
            }

            // ------------------------------------------------
            // Example: FunctionDeclaration
            // ------------------------------------------------
            case TypedTreeNode.FunctionDeclaration fd -> {
                TypedTreeNode.TypedIdentifier newTypedIdentifier = (TypedTreeNode.TypedIdentifier) apply(fd.name());

                var newParams = fd.parameters().stream()
                        .map(this::apply)
                        .map(TypedTreeNode.Parameter.class::cast)
                        .toList();

                TypedTreeNode.Expression newBody = (TypedTreeNode.Expression) apply(fd.body());
                yield new TypedTreeNode.FunctionDeclaration(
                        newTypedIdentifier,
                        newParams,
                        newBody
                );
            }

            // ------------------------------------------------
            // Example: Parameter
            // ------------------------------------------------
            case TypedTreeNode.Parameter param -> {
                TypedTreeNode.TypedIdentifier typedIdentifier = (TypedTreeNode.TypedIdentifier) apply(param.identifier());
                yield new TypedTreeNode.Parameter(typedIdentifier);
            }

            // ------------------------------------------------
            // Example: TypedIdentifier
            // ------------------------------------------------
            case TypedTreeNode.TypedIdentifier ti -> {
                // Possibly no children to transform, so just rebuild with newType
                final var newType = unifier.applySubstitution(ti.type());
                yield new TypedTreeNode.TypedIdentifier(ti.identifier(), newType);
            }

            // ------------------------------------------------
            // Example: Expression subtypes
            // ------------------------------------------------
            case TypedTreeNode.Expression expr -> switch (expr) {
                // IntegerLiteral
                case TypedTreeNode.IntegerLiteral intLit -> intLit;

                // StringLiteral
                case TypedTreeNode.StringLiteral strLit -> strLit;

                // PrintExpression
                case TypedTreeNode.PrintExpression printExpr -> {
                    TypedTreeNode.Expression newInner = (TypedTreeNode.Expression) apply(printExpr.expression());
                    yield new TypedTreeNode.PrintExpression(newInner);
                }

                // FunctionCall
                case TypedTreeNode.FunctionCall funcCall -> {
                    TypedTreeNode.TypedIdentifier typedIdentifier = (TypedTreeNode.TypedIdentifier) apply(funcCall.name());
                    var newArguments = funcCall.arguments().stream()
                            .map(this::apply)
                            .map(TypedTreeNode.Expression.class::cast)
                            .toList();
                    yield new TypedTreeNode.FunctionCall(
                            typedIdentifier,
                            newArguments
                    );
                }

                // ArithmeticExpression
                case ArithmeticExpression aexpr -> switch (aexpr) {
                    case TypedTreeNode.Operand operand -> {
                        var newType = unifier.applySubstitution(operand.type());
                        var newLeft = apply(operand.left());
                        yield new TypedTreeNode.Operand(newType, (TypedTreeNode.Term) newLeft);
                    }
                    case BinaryArithmeticExpression bae -> {
                        var newType = unifier.applySubstitution(bae.type());
                        var newLeft = apply(bae.left());
                        var newRight = apply(bae.right());
                        yield new BinaryArithmeticExpression(
                                newType,
                                (TypedTreeNode.Term) newLeft,
                                (TypedTreeNode.ArithmeticExpression) newRight,
                                bae.operation()
                        );
                    }
                };
                case NotImplemented notImplemented -> notImplemented;
                case ParenthesizedExpression parenthesizedExpression ->
                        new ParenthesizedExpression((ArithmeticExpression) apply(parenthesizedExpression.arithmeticExpression()));
                case TypedTreeNode.Term term -> switch (term) {
                    case TypedTreeNode.SimpleTerm simpleTerm -> {
                        final var newType = unifier.applySubstitution(simpleTerm.type());
                        final var newLeft = apply(simpleTerm.left());
                        yield new TypedTreeNode.SimpleTerm(newType, (TypedTreeNode.Factor) newLeft);
                    }
                    case TypedTreeNode.BinaryTerm binaryTerm -> {
                        final var newType = unifier.applySubstitution(binaryTerm.type());
                        final var newLeft = apply(binaryTerm.left());
                        final var newRight = apply(binaryTerm.right());
                        yield new TypedTreeNode.BinaryTerm(
                                newType,
                                (TypedTreeNode.Factor) newLeft,
                                (TypedTreeNode.Term) newRight,
                                binaryTerm.operation()
                        );
                    }
                };
            };
            case TypedTreeNode.NamespaceDeclaration namespaceDeclaration -> namespaceDeclaration;
        };
    }
}

File: src/main/scala/com/github/tammo/frontend/type/SubstitutionApplier.scala
package com.github.tammo.frontend.`type`

import com.github.tammo.frontend.`type`.TypedTree.*

object SubstitutionApplier {

  def applySubstitutions(
      typedTree: TypedTree,
      apply: Type => Type
  ): TypedTree = typedTree match
    case compilationUnit: CompilationUnit =>
      applySubstitutionToCompilationUnit(compilationUnit, apply)
    case namespaceDeclaration: NamespaceDeclaration =>
      namespaceDeclaration
    case classDeclaration: ClassDeclaration =>
      applySubstitutionToClassDeclaration(classDeclaration, apply)
    case declaration: EffectDeclaration =>
      applySubstitutionToEffectDeclaration(declaration, apply)
    case declaration: FunctionDeclaration =>
      applySubstitutionToFunctionDeclaration(declaration, apply)
    case parameter: Parameter => applySubstitutionToParameter(parameter, apply)
    case identifier: TypedIdentifier =>
      applySubstitutionToTypedIdentifier(identifier, apply)
    case NotImplemented => NotImplemented
    case expression: Expression =>
      applySubstitutionToExpression(expression, apply)

  private def applySubstitutionToCompilationUnit(
      compilationUnit: CompilationUnit,
      apply: Type => Type
  ): CompilationUnit = CompilationUnit(
    compilationUnit.namespace,
    applySubstitutionToClassDeclaration(
      compilationUnit.classDeclaration,
      apply
    ),
    compilationUnit.span
  )

  private def applySubstitutionToClassDeclaration(
      classDeclaration: ClassDeclaration,
      apply: Type => Type
  ): ClassDeclaration = ClassDeclaration(
    classDeclaration.name,
    classDeclaration.effects.map(
      applySubstitutionToEffectDeclaration(_, apply)
    ),
    classDeclaration.functions.map(
      applySubstitutionToFunctionDeclaration(_, apply)
    ),
    classDeclaration.span
  )

  private def applySubstitutionToEffectDeclaration(
      effectDeclaration: EffectDeclaration,
      apply: Type => Type
  ): EffectDeclaration = EffectDeclaration(
    applySubstitutionToTypedIdentifier(effectDeclaration.identifier, apply),
    effectDeclaration.parameters.map(applySubstitutionToParameter(_, apply)),
    applySubstitutionToExpression(effectDeclaration.body, apply),
    effectDeclaration.span
  )

  private def applySubstitutionToFunctionDeclaration(
      functionDeclaration: FunctionDeclaration,
      apply: Type => Type
  ): FunctionDeclaration = FunctionDeclaration(
    applySubstitutionToTypedIdentifier(functionDeclaration.identifier, apply),
    functionDeclaration.parameters.map(applySubstitutionToParameter(_, apply)),
    applySubstitutionToExpression(functionDeclaration.body, apply),
    functionDeclaration.span
  )

  private def applySubstitutionToTypedIdentifier(
      typedIdentifier: TypedIdentifier,
      apply: Type => Type
  ): TypedIdentifier = TypedIdentifier(
    typedIdentifier.name,
    apply(typedIdentifier.`type`),
    typedIdentifier.span
  )

  private def applySubstitutionToParameter(
      parameter: Parameter,
      apply: Type => Type
  ): Parameter = Parameter(
    applySubstitutionToTypedIdentifier(parameter.identifier, apply),
    parameter.span
  )

  private def applySubstitutionToExpression(
      expression: Expression,
      apply: Type => Type
  ): Expression = expression match
    case PrintExpression(expression, span) =>
      PrintExpression(applySubstitutionToExpression(expression, apply), span)
    case ParenthesizedExpression(expression, span) =>
      ParenthesizedExpression(
        applySubstitutionToExpression(expression, apply),
        span
      )
    case literal: StringLiteral => literal
    case application: FunctionApplication =>
      applySubstitutionToFunctionApplication(application, apply)
    case expression: ArithmeticExpression =>
      applySubstitutionToArithmeticExpression(expression, apply)
    case term: Term     => applySubstitutionToTerm(term, apply)
    case factor: Factor => applySubstitutionToFactor(factor)

  private def applySubstitutionToFunctionApplication(
      functionApplication: FunctionApplication,
      apply: Type => Type
  ): FunctionApplication = FunctionApplication(
    applySubstitutionToTypedIdentifier(functionApplication.identifier, apply),
    functionApplication.arguments.map(applySubstitutionToExpression(_, apply)),
    functionApplication.span
  )

  private def applySubstitutionToArithmeticExpression(
      expression: ArithmeticExpression,
      apply: Type => Type
  ): ArithmeticExpression = expression match
    case operand: Operand =>
      Operand(
        apply(operand.`type`),
        applySubstitutionToTerm(operand.left, apply),
        operand.span
      )
    case expression: BinaryArithmeticExpression =>
      BinaryArithmeticExpression(
        apply(expression.`type`),
        applySubstitutionToTerm(expression.left, apply),
        applySubstitutionToExpression(expression.right, apply),
        expression.operator,
        expression.span
      )

  private def applySubstitutionToTerm(
      term: Term,
      apply: Type => Type
  ): Term = term match
    case term: UnaryTerm =>
      UnaryTerm(
        apply(term.`type`),
        applySubstitutionToFactor(term.left),
        term.span
      )
    case term: BinaryTerm =>
      BinaryTerm(
        apply(term.`type`),
        applySubstitutionToFactor(term.left),
        applySubstitutionToExpression(term.right, apply),
        term.operator,
        term.span
      )

  private def applySubstitutionToFactor(factor: Factor): Factor = factor match
    case literal: IntLiteral => literal

}
